lib/pf/Switch/Ruckus/SmartZone.pm

```diff
@@ -204,39 +208,61 @@ sub deauthenticateMacWebservices {
     my $logger = $self->logger;
     my $controllerIp = $self->{_controllerIp} || $self->{_ip};
     my $webservicesPassword = $self->{_wsPwd};
+    
     my $ucmac = uc $mac;
     my $ip = pf::ip4log::mac2ip($mac);
     my $node_info = node_view($mac);
     my $payload;
-    if($node_info->{status} eq "unreg" || security_event_count_reevaluate_access($mac)) {
-        $payload = encode_json({
-         "Vendor"=> "ruckus",
-         "RequestPassword"=> $webservicesPassword,
-         "APIVersion"=> "1.0",
-         "RequestCategory"=> "UserOnlineControl",
-         "RequestType"=> "Logout",
-         "UE-IP"=> $ip,
-         "UE-MAC"=> $ucmac
-        });
+
+    my %baseCommand=(
+        "Vendor"=> "ruckus",
+        "RequestPassword"=> $webservicesPassword,
+        "APIVersion"=> "1.0",
+        "RequestCategory"=> "UserOnlineControl",
+        "UE-IP"=> $ip,
+        "UE-MAC"=> $ucmac
+    );
+
+    # If a webservice username is defined, add the key/value to the hash so that it appears on the json.
+    # Otherwise, the "RequestUserName" field should not exist at all.
+    # RequestUserName is used in "managed Service Provider" domains in SmartZone HighScale
+    # See here: https://docs.commscope.com/bundle/sz-510-hotspot-wispr-guide-sz300vsz/page/GUID-160E59E5-5816-4618-B0D4-091FAA9AD49C.html#
+
+    if (defined($self->{'_wsUser'}) && $self->{'_wsUser'} ne ''){
+        $baseCommand{"RequestUserName"}=$self->{'_wsUser'};
     }
-    else {
-        $payload = encode_json({
-         "Vendor"=> "ruckus",
-         "RequestPassword"=> $webservicesPassword,
-         "APIVersion"=> "1.0",
-         "RequestCategory"=> "UserOnlineControl",
-         "RequestType"=> "Login",
-         "UE-IP"=> $ip,
-         "UE-MAC"=> $ucmac,
-         "UE-Proxy"=> "0",
-         "UE-Username"=> $ucmac,
-         "UE-Password"=> $ucmac
-        });
+
+    if ( $node_info->{last_connection_type} eq $connection_type_to_str{$WIRELESS_MAC_AUTH} ){
+        # For Ruckus smartzone, if using non-proxy RADIUS (MAC auth), the CoA/disconnect is via web services.
+        # We need to invoke the "Disconnect" method every time we want the user to change status. For mac-auth, "Login" or "Logout" will not trigger a vlan change / reconnection
+        # So basically, the user needs to configure the deauth method as http/https to force it to use "deauthenticateMacWebservices"
+        # and once here, we check if the connection was WIRELESS_MAC_AUTH and trigger the logout webservice call
+        # "Disconnect" also forces the ap to disociate the client which is what we want for unreg/role revaluation.
+        $baseCommand{"RequestType"}="Disconnect";
+    } elsif( $node_info->{status} eq "unreg" || security_event_count_reevaluate_access($mac) ){
+        $baseCommand{"RequestType"}="Logout";
```

Comment:
I kept the previous' implementation call of "logout" even if im not 100% sure it works (not sure it will cause the client to request a new IP). I havent done any testing using the WISPr / hotspot2.0 mechanism so I didnt want to change the existing implementation.